### PR TITLE
[dashboard] update preferences page for desktop IDE's

### LINKED ISSUE #7568: Update preferences page for desktop IDE's
On the preferences page for IDE's, currently the user is presented with the default browser editor, and the desktop one is behind a toggle. This update is to make the desktop IDE's more prominent in the UI, so they don't appear hierarchically as a second-class citizen. 

**Note:** there are many more changes required to make the desktop IDE UX smoother https://github.com/gitpod-io/gitpod/issues/6602, this is a small incremental change to make steps towards a better desktop UX. 

**Main proposed changes:** 
* Remove toggle for browser/desktop so all IDE's/editors are viewable by default
* Introduces a "none" preference for desktop IDE's (instead of the toggle)

|Before|After|
|--|--|
|image|image|

Link to [Figma design](https://www.figma.com/file/zgiOqtQm3AH3HOCSnd4jK2/Editor-preferences-page-%237568?node-id=0%3A1)

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

update preferences page for desktop IDE's. Link to [Figma design](https://www.figma.com/file/zgiOqtQm3AH3HOCSnd4jK2/Editor-preferences-page-%237568?node-id=0%3A1)

## Related Issue(s)

Fixes #7568

## How to test

1. Page display like Figma said
2. Socket trace event works like before

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **laushinka** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @laushinka` in a comment when ready.

Associated issue: *#7568*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by akosyakov** [comment]
@mustard-mh btw you can convert a PR to draft while working on it

**Comment by mustard-mh** [comment]
Fixed @gtsiolis

**Comment by gtsiolis** [comment]
Thanks, @mustard-mh!

Looking at this now! 👀

File: components/dashboard/src/settings/Preferences.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/Preferences.tsx:134]: **praise**: Thanks so much for wrapping the options here, @mustard-mh! 🌟

See also :a: [relevant discussion](https://gitpod.slack.com/archives/C02CBKZEVS8/p1638795733242600?thread_ts=1638366112.229500&cid=C02CBKZEVS8) (internal) and :b: https://github.com/gitpod-io/gitpod/pull/7031#discussion_r766124754 for future reference.

File: components/dashboard/src/settings/Preferences.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/Preferences.tsx:24]: **suggestion:** Pasting below a possible logo option we could use here.

```svg
<svg width="64" height="64" fill="none" xmlns="http://www.w3.org/2000/svg"><g clip-path="url(#a)"><path fill-rule="evenodd" clip-rule="evenodd" d="M50.102 13.898c-9.997-9.997-26.206-9.997-36.204 0-9.997 9.997-9.997 26.207 0 36.204s26.207 9.997 36.204 0 9.997-26.206 0-36.204ZM36.526 41.051a3.2 3.2 0 0 0 4.525-4.526L27.474 22.95a3.2 3.2 0 0 0-4.525 4.525l13.576 13.577Z" fill="#D6D3D1"/></g><defs><clipPath id="a"><path fill="#fff" d="M0 0h64v64H0z"/></clipPath></defs></svg>
```

| None option logo |
|-|
| IDE-options |

File: components/dashboard/src/settings/Preferences.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/Preferences.tsx:213]: **suggestion**: We can probably safely remove ther left margin here.

```suggestion
        {option.label ? <div className={`font-semibold text-sm ${selected ? 'text-green-500' : 'text-gray-500 dark:text-gray-400'} uppercase mt-2 px-3 py-1 self-center`}>{option.label}</div> : <></>}
```

**Review by gtsiolis — APPROVED** [review]
Thanks, @mustard-mh! 🔮 

Left some minor non-blocking comments below, regarding UX and copy changes.

I'd defer to @loujaybee for decicing what to ship in iteration. 🏓 

Approving to unblock merging but holding to let @loujaybee decide on the minor suggestions below.

/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 1a4879c500ee2aa318a00fc6ac765e52a97b6ab2</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7653#pullrequestreview-855377416" title="Approved">gtsiolis</a>*

Associated issue: *#7568*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

File: components/dashboard/src/settings/Preferences.tsx
Comment (gtsiolis) [reply components/dashboard/src/settings/Preferences.tsx:24]: @mustard-mh I think so, but not entirely sure. See https://github.com/gitpod-io/gitpod/pull/7141 and https://github.com/gitpod-io/gitpod/issues/6955. Cc @iQQBot @akosyakov

Comment (akosyakov) [reply components/dashboard/src/settings/Preferences.tsx:24]: It is fine here as well. I don't think we are going to change it often. We put logos on ide proxy to add new IDEs without involving meta team. General changes of preferences UI are deployed by meta.

Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx]: Already changed, please review again

Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx]: image
image
SVGs looks incongruous (none got a padding?)

File: components/dashboard/src/settings/Preferences.tsx
Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx:134]: Another PR later will be good (

File: components/dashboard/src/settings/Preferences.tsx
Comment (gtsiolis) [reply components/dashboard/src/settings/Preferences.tsx]: Thanks for noticing, @mustard-mh! ✨ 

The :a: extra padding was added to make this icon less prominent and not stand out as much as the rest of the other editor icons as well as :b: provide some visual balance next to other editor icons (logos) which don't use the whole icon bounding box. If you notice the overall size of all of them still remains the same (64x64px) in order to maintain consistent layout.

| Using a different live area (PROPOSED) | Using bounding box size (ALTERNATIVE) |
|-|-|
| Preferences | Preferences-1 |

Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx]: It looks more stand out in dark mode
image

Comment (gtsiolis) [reply components/dashboard/src/settings/Preferences.tsx]: Nice catch, @mustard-mh! ⚾ 

I'd be certain the grayscale filter could help but not in this case. 💭 

We could use different versions with different colors but a boring (simple) solution here could be to just use a different color for both. Let's go with that since we're probably going to update this section anyway.

```svg
<svg width="64" height="64" fill="none" xmlns="http://www.w3.org/2000/svg"><g clip-path="url(#a)"><path fill-rule="evenodd" clip-rule="evenodd" d="M50.102 13.898c-9.998-9.997-26.207-9.997-36.204 0s-9.997 26.207 0 36.204 26.206 9.997 36.204 0c9.997-9.997 9.997-26.206 0-36.204ZM36.525 41.051a3.2 3.2 0 0 0 4.526-4.526L27.474 22.95a3.2 3.2 0 1 0-4.525 4.525l13.576 13.577Z" fill="#A8A29E"/></g><defs><clipPath id="a"><path fill="#fff" d="M0 0h64v64H0z"/></clipPath></defs></svg>
```

| Light | Dark |
|-|-|
| desktop-light | desktop-dark |

Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx]: let it go, since it is not a big problem

Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx]: hey @gtsiolis 👋, just make dark logo visible and remove some code

**Comment by mustard-mh** [comment]
> Thanks, @mustard-mh! 🔮
> 
> Left some minor non-blocking comments below, regarding UX and copy changes.
> 
> I'd defer to @loujaybee for deciding what to ship in this iteration. 🏓
> 
> Approving to unblock merging but holding to let @loujaybee decide on the minor suggestions below.
> 
> /hold

Please review again or more further suggestions 🙏  ? or block for further design?
cc @loujaybee

**Comment by gtsiolis** [comment]
UX [LGTM](https://en.wiktionary.org/wiki/LGTM), @mustard-mh!

@jankeromnes could you :a: take a closer look at the code changes and :b: check if keeping the SVG icons in `components/dashboard/public/image` is fine here? 🏀

**Comment by jankeromnes** [comment]
Thanks! I can take a look. 👀

File: components/dashboard/public/image/none-dark.svg
Comment (jankeromnes) [inline components/dashboard/public/image/none-dark.svg:1]: **Praise**: Thanks for optimizing the SVGs! 🙏

Not super important, but we already have the same icon (rotated) over here:

https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/icons/StatusCanceled.svg

No need to fix this here, but in general I think it would be good to make the location & naming of our icons more consistent.

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:39]: Whoa, I had to read [MDN: Nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) again in order to understand this. 😅

**Suggestion**: I think using the nullish coalescing operator is not necessary here, as a simple `||` can do the job (and is much more common and readable):

```suggestion
        const additionalData = user?.additionalData || {};
        const settings = additionalData.ideSettings || {};
```

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:37]: **Suggestion**: No need to make this mutable:

```suggestion
        const useDesktopIde = defaultDesktopIde !== DesktopNoneId;
```

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:51]: **Question**: Why is change needed? (Doesn't the console already log uncaught Promise errors?)

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:58]: **Praise**: Thank you for de-duplicating this code. 👍 ✨

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:134]: This isn't needed, right? (I think `selected` above will already be correct when `id === DesktopNoneId`)

```suggestion
```

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:135]: **Thought**: When possible, I prefer the simplicity of using only one image in both Light and Dark themes. Here, having two different images seems a bit too complicated for my taste.

@gtsiolis Do you think we could get by with only one `None` image in both Light and Dark themes? (If we can, we would have only one SVG image, and we could remove all the extra `isDark` logic added here.)

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:206]: All `IDEOption` should always have a `logo`, right? If so, maybe you could undo the extra `option.logo` check here?

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:208]: **Praise**: Wow, many thanks for removing this very unsatisfying left margin. 😌

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:107]: `browserIdeOptions` is already checked on the line above, no need to check a second time:

```suggestion
```

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:119]: **Suggestion**: I think this could be simplified:

```suggestion
                    {ideOptions.options[defaultIde]?.notes &&
```

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:141]: **Suggestion**: I think this could be simplified:

```suggestion
                    {ideOptions.options[defaultDesktopIde]?.notes &&
```

**Review by jankeromnes — COMMENTED** [review]
Many thanks @mustard-mh for improving the IDE selection code! 🙏

I read the code and the improvements look very good to me overall. 👍 Still, I had a few questions/suggestions, which I added in-line.

Could you please take a look at my comments, and let me know if you agree or disagree with them?

Also, the preview environment seems broken, so I couldn't test this PR, but that's okay since @gtsiolis has already done that. ✅

File: components/dashboard/src/settings/Preferences.tsx
Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx:51]: Mostly because that it will crash in Nodejs v16, so it's good to add this for `Promise`

File: components/dashboard/src/settings/Preferences.tsx
Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx:39]: Nullish coalescing operator `??` is for disambiguation, we should give preference to `??`
Here we use `??`, means that if `user?.additionalData` is null we got a default value `{}`

eg:
```
// data = { boolValue: false }
let value = data?.boolValue || true     // true
let value = data?.boolValue ?? true.  // false
```

File: components/dashboard/src/settings/Preferences.tsx
Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx]: yes

File: components/dashboard/public/image/none-dark.svg
Comment (mustard-mh) [reply components/dashboard/public/image/none-dark.svg]: I have moved and renamed SVGs

**Comment by mustard-mh** [comment]
Also wondering if there has any formatter for dashboard? @jankeromnes

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [reply components/dashboard/src/settings/Preferences.tsx:39]: Indeed, this is correct for number and booleans, but here `additionalData` and `ideSettings` should always be objects.

The only case where `??` and `||` have a different behavior is if `additionalData` has a non-object, non-null value, e.g. `false` or `''`. These cases are theoretically impossible, but even if they happen, I think we'd still prefer setting the default value of `{}` instead of keeping the bad value `false` or `''`.

My point is that `??` is a rare feature, which might hurt the readability and maintainability of the dashboard code.

However, I don't have a very strong opinion, so if you insist on keeping the `??` here, I won't block your valuable Pull Request for that. 😇

File: components/dashboard/src/settings/Preferences.tsx
Comment (gtsiolis) [reply components/dashboard/src/settings/Preferences.tsx]: @jankeromnes Yeah, we could go with a single SVG icon as described in https://github.com/gitpod-io/gitpod/pull/7653#discussion_r786800052. Later on, we could fine tune this and either require two SVG icons again or apply proper style filters on top of it.

**Comment by jankeromnes** [comment]
> Also wondering if there has any formatter for dashboard? @jankeromnes

Good question. I think when you do `F1 > Format Document`, it does something (maybe with Prettier?) but I'm not sure where this is configured.

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:42]: **Thought**: Here, you might be setting `"none"` as the `defaultDesktopIde` in `user.additionalData.ideSettings`.

It seems a little strange that `user.additionalData.ideSettings.defaultDesktopIde` can have a value that is not present in [the configmap](https://github.com/gitpod-io/gitpod/blob/4bdcf8c8f203795c2951a19940e3151a57795f1e/chart/templates/server-ide-configmap.yaml#L61-L103).

Maybe, just to be safe, we should add the `none` option in the configmap, or set `defaultDesktopIde` to `undefined` instead of `"none"`?

File: components/dashboard/src/settings/Preferences.tsx
Comment (jankeromnes) [inline components/dashboard/src/settings/Preferences.tsx:61]: **Bug**: As a brand new User, I see this in my Preferences page (if I don't change anything):

Screenshot 2022-01-20 at 13 04 48

You can see that `VS Code` is selected as my default Desktop IDE.

However, if I start a workspace, it goes straight to Browser IDE (it doesn't propose to use my preferred Desktop IDE). To make it propose a Desktop IDE, I need to first select `None`, and then `VS Code` again.

Wouldn't it make more sense to show `None` as selected by default, when `useDesktopIde` is false?

```suggestion
    const [defaultDesktopIde, setDefaultDesktopIde] = useState<string>(user?.additionalData?.ideSettings?.useDesktopIde && user?.additionalData?.ideSettings?.defaultDesktopIde || DesktopNoneId);
```

**Review by jankeromnes — APPROVED** [review]
Many thanks for making all these fixes so quickly! :pray:

The code now looks good to me, so I'll approve it.

However, I've noticed a few small problems on second read, so I've added a few more comments and I'll add a hold on this PR so that it doesn't get merged automatically. Please take a look at my comments, and decide whether to address them or not.

Alternatively, if this PR is time-critical, please feel free to remove/cancel the hold, so that this PR gets merged, and to open follow-up issues to address my comments later.

/lgtm
/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: cfb362e2a950c4a9af750fd3d28b9bf191d0f624</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7653#pullrequestreview-855377416" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7653#pullrequestreview-858141775" title="LGTM">jankeromnes</a>*

Associated issue: *#7568*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis,jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

File: components/dashboard/src/settings/Preferences.tsx
Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx:40]: Yes, I insist my option. It's fine here to use `||`, but with `??` we don't need to think about, what `additionalData` is, is it a bad zero value for this expression or not?

File: components/dashboard/src/settings/Preferences.tsx
Comment (mustard-mh) [reply components/dashboard/src/settings/Preferences.tsx]: Thanks 🙂

**Comment by mustard-mh** [comment]
> > Also wondering if there has any formatter for dashboard? @jankeromnes
> 
> Good question. I think when you do `F1 > Format Document`, it does something (maybe with Prettier?) but I'm not sure where this is configured.

I think it's better to put some lint or format config in dashboard
For me, in this project, I often forget convert `let` to `const`, add `;` in the trailing and some incorrectly indent 😅

**Comment by akosyakov** [comment]
@mustard-mh don't forget to unhold if you think it is good to go

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 0106c07905f15ef03b9b8c82bea54a113a33784e</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7653#issuecomment-1018277242" title="LGTM">akosyakov</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7653#pullrequestreview-855377416" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7653#pullrequestreview-858141775" title="LGTM">jankeromnes</a>*

Associated issue: *#7568*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis,jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
